src/main/ruby/truffleruby/core/struct.rb

```diff
@@ -337,6 +337,33 @@ def to_a
     _attrs.map { |var| Primitive.object_hidden_var_get(self, var) }
   end
   alias_method :values, :to_a
+  alias_method :deconstruct, :to_a
+
+  def deconstruct_keys(keys)
+    return to_h if Primitive.nil?(keys)
+    raise TypeError, "wrong argument type #{Truffle::Type.object_class(keys)} (expected Array or nil)" unless keys.is_a?(Array)
```

Comment:
Fixed. For spectators this is because `coerce_to` wraps up this common pattern.

Reply:
This isn’t right (yet) because [the spec](https://github.com/ruby/ruby/blob/c2a60fec2f79c05bdb865c143b6ad8eddfc6cc36/spec/ruby/core/struct/deconstruct_keys_spec.rb#L64) expects a `TypeError` whose message includes “`expected Array or nil`”, whereas `Truffle::Type.coerce_to` raises a `TypeError` with message “`Coercion error: "x".to_ary => Array failed`”. Is there a way for us to control or change this message, or is it better to catch & rethrow?

Reply:
It seems in this case no coercion happens (which is exception, coercion is far more frequent), so the original code was fine and used the special error message for this case.
```
[7] pry(main)> o=Object.new
=> #<Object:0x0000557ca01bdc90>
[8] pry(main)> def o.to_a; [:a]; end
=> :to_a
[9] pry(main)> Struct.new(:a,:b).new(1,2).deconstruct_keys(o)
TypeError: wrong argument type Object (expected Array or nil)
from (pry):9:in `deconstruct_keys'
[10] pry(main)> def o.to_ary; [:a]; end
=> :to_ary
[11] pry(main)> Struct.new(:a,:b).new(1,2).deconstruct_keys(o)
TypeError: wrong argument type Object (expected Array or nil)
from (pry):11:in `deconstruct_keys'
```

So my suggestion was incorrect, sorry about that.

Reply:
No problem, thanks for confirming. I’ll back out this part of the change.